pkg/distributor/distributor.go

```diff
@@ -629,6 +643,7 @@ func (d *Distributor) wrapPushWithMiddlewares(next push.Func) push.Func {
 	// To guarantee that, middleware functions will be called in reversed order, wrapping the
 	// result from previous call.
 	middlewares = append(middlewares, d.instanceLimitsMiddleware) // should run first
+	middlewares = append(middlewares, d.prePushUserRequestMiddleware)
```

Comment:
the `instanceLimitsMiddleware` could already reject some requests, shouldn't those be counted too?

Reply:
I'm not sure. We don't increment per-tenant metrics for anything else before checking the instance limits.

The way the other per-tenant limits work is:
* The total metric is "up to this number of X may be accepted"
* The rejected metric is "this number of X was rejected or dropped for reason Y"
* The received metric is "this number made it all the way to a distributor and will be sent to an ingester"

I'm having trouble fitting the instance limits into that model. We'd increase the "up to this number of X may be accepted" metric but then they'd all be rejected due to something the user has no control over.

Reply:
Additionally, hitting the instance limits results in 500 errors which will be retried by the user. If we didn't run that middleware first, we're potentially double counting user samples/exemplars/requests